include/exadg/compressible_navier_stokes/user_interface/implement_get_application.h

```diff
@@ -0,0 +1,43 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2021 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef INCLUDE_EXADG_COMPRESSIBLE_NAVIER_STOKES_USER_INTERFACE_IMPLEMENT_GET_APPLICATION_H_
+#define INCLUDE_EXADG_COMPRESSIBLE_NAVIER_STOKES_USER_INTERFACE_IMPLEMENT_GET_APPLICATION_H_
+
+#include <exadg/compressible_navier_stokes/user_interface/application_base.h>
+
+namespace ExaDG
+{
+namespace CompNS
+{
+template<int dim, typename Number>
+std::shared_ptr<ApplicationBase<dim, Number>>
+get_application(std::string input_file, MPI_Comm const & comm)
+{
+  return std::make_shared<Application<dim, Number>>(input_file, comm);
+}
+
+} // namespace CompNS
+} // namespace ExaDG
```

Comment:
Could you explain the purpose of these `declare_get_application.h` and `implement_get_application.h`. Is the only purpose so that one does not have to write `std::make_shared<Application<dim, Number>>(input_file, comm)` directly? But one has to include this fill in all applications.

Reply:
Yes, I would have to touch this function for all the applications whenever the interface changes. So I decided to implement it only once and include it in all applications (technically, this is possible since we have separate executables for each application and since each application can therefore have the same name `Application`, which is different from the "ApplicationSelector" concept with different name for each application which we had some time ago). I know the code is only a single line of implementation, but I came to the conclusion that being more strict in terms of "don't repeat yourself" makes sense here.

Reply:
I won't block this change, since the general structure of `get_application()` is not new.

However, I don't see how this helps to reduce code duplication. Currently, there are 7 `declare_get_application.h` and 7 `implement_get_application.h` files and `implement_get_application.h` has to be included in all applications. My question is why does someone has to return a shared pointer of the application. Why isn't it enough to let the user provide an application class and let the solver/drive create a shared pointer of it.

Reply:
7 is not much if you count - in comparison - the number of all applications (which I would need to change due to the additional mpi_comm parameter introduced in this PR).

But the driver/solver does not know about the concrete application. Principally, one could let the driver instantiate the application, but this would force the user to always name a particular application `Application`. note that `solver.h` (and also `driver.h`) only contains a forward declaration with `ApplicationBase`! My suggestion is much more flexible, as it allows to name the application whatever the user likes best, e.g., `MyFancyApplication`. In this case, the user than just has to instantiate the application manually instead of including the new `implement_get_application()`. I argue that such a user knows what to do. However, what I want is to minimize the effort for unexperienced users. Adding another switch-branch to the `ApplicationSelector` as we had some time ago is already too complex for a new user that just wants to experiment with ExaDG in my opinion.

Reply:
Let's go with the current approach. Maybe there will be cases that one want to have multiple fluid applications and in that case they need different names or have to be in different namespaces.